contrib/IECoreArnold/src/IECoreArnold/MeshAlgo.cpp

```diff
@@ -84,121 +84,44 @@ const T *variableData( const PrimitiveVariableMap &variables, const std::string
-//
-// All this plays poorly with Arnold, because it stores UVs sanely, as an array of 2d vectors with
-// an indices array to index into them per face-vertex. So this function converts the Cortex
-// representation into the Arnold form, so that Arnold can use the connectivity to correctly
-// interpolate the uvs.
-//
-// We should really improve the Cortex representation as follows :
-//
-// - Use V2fVectorData for UVs.
-// - Add an optional IntVectorData indices array into the PrimitiveVariable class to store
-//   indices into the UV array, and associating the indices directly with the data.
-// - Use the above to avoid the duplication of UVs when storing FaceVarying UVs, so we can
-//   do a straightforward conversion to Arnold, and only do the dumb duplication on conversion
-///  to RenderMan.
-// - Perhaps also add a role enum to PrimitiveVariable, so we can distinguish between UVs and
-///  things that just happen to hold V2fVectorData.
-void convertIndexedUVSet( const std::string &setName, PrimitiveVariableMap &variables, AtNode *node )
-{
-	const std::string sName = setName == "" ? "s" : setName + "_s";
-	const std::string tName = setName == "" ? "t" : setName + "_t";
-	const std::string indicesName = setName == "" ? "stIndices" : setName + "Indices";
-
-	const FloatVectorData *sData = variableData<FloatVectorData>( variables, sName, PrimitiveVariable::FaceVarying );
-	const FloatVectorData *tData = variableData<FloatVectorData>( variables, tName, PrimitiveVariable::FaceVarying );
-	const IntVectorData *indicesData = variableData<IntVectorData>( variables, indicesName, PrimitiveVariable::FaceVarying );
-
-	if( !( sData && tData && indicesData ) )
-	{
-		return;
-	}
-
-	const vector<float> &s = sData->readable();
-	const vector<float> &t = tData->readable();
-	const vector<int> &indices = indicesData->readable();
-
-	int numUVs = 1 + *max_element( indices.begin(), indices.end() );
-	AtArray *uvsArray = AiArrayAllocate( numUVs, 1, AI_TYPE_POINT2 );
-	AtArray *indicesArray = AiArrayAllocate( indices.size(), 1, AI_TYPE_UINT );
-
-	for( size_t i = 0, e = indices.size(); i < e; ++i )
-	{
-		AtPoint2 uv = { s[i], t[i] };
-		AiArraySetPnt2( uvsArray, indices[i], uv );
-		AiArraySetUInt( indicesArray, i, indices[i] );
-	}
-
-	if( setName == "" )
-	{
-		AiNodeSetArray( node, "uvlist", uvsArray );
-		AiNodeSetArray( node, "uvidxs", indicesArray );
-	}
-	else
-	{
-		AiNodeDeclare( node, setName.c_str(), "indexed POINT2" );
-		AiNodeSetArray( node, setName.c_str(), uvsArray );
-		AiNodeSetArray( node, (setName + "idxs").c_str(), indicesArray );
-	}
-
-	variables.erase( sName );
-	variables.erase( tName );
-	variables.erase( indicesName );
-}
-
-// Version of the above for when we have no indices available.
-void convertUVSet( const std::string &setName, PrimitiveVariableMap &variables, const vector<int> &vertexIds, AtNode *node )
+void convertUVSet( PrimitiveVariableMap::const_iterator uvIt, const vector<int> &vertexIds, AtNode *node )
```

Comment:
Iterator seems a strange choice of type, since we don't do any iteration - perhaps `const PrimitiveVariable &`? Ah, I see, we need the name as well as the PrimitiveVariable - still feels a bit odd though...

Reply:
Its anonymous namespace, so I figured the signature wasn't too important to cleanup. I could pass name and variable if you prefer, but it does seem a bit redundant. Also, I'm calling `erase` at the end, which I guess could be lifted out of the function back to the call site.

Reply:
> Its anonymous namespace, so I figured the signature wasn't too important to cleanup.

Yeah, not the most important thing for sure. I just felt that passing an iterator implies doing something that needs the iterator, which isn't the case here.

> I could pass name and variable if you prefer, but it does seem a bit redundant.

Or a `const PrimitiveVariableMap::value_type &`, which is what the iterator points to. I do think `name, value` is more readable than `first, second` though.

> Also, I'm calling erase at the end, which I guess could be lifted out of the function back to the call site.

I'm not seeing that - you'd need the container to call erase, and it's not being passed as an argument. The erase appears to already be lifted out into `convertCommon()`. This confusion seems like the best argument for changing the type - there can't be any confusion about the responsibility if `convertUVSet()` has no iterator...

Reply:
Addressed in 214ec0781bc75d1f48bd413abcd407f9cb56e69c